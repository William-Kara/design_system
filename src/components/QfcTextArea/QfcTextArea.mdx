import { ArgTypes, Meta } from '@storybook/blocks';
import { Story } from '@storybook/addon-docs';
import QfcTextArea from '@/components/QfcTextArea/QfcTextArea.vue';

<Meta title="Elements/Inputs/TextArea" />

# Definition & Usage

A customizable textarea component.

## Props

The following props are available for the QfcTextArea component:

<ArgTypes of={QfcTextArea} />

When the isRequired prop is set to true, an asterisk appears to the right of the label.

The input field become red when the errorMessage prop is set.

# Declinaison

## Basic

<Story name="Basic" id="elements-inputs-textarea--primary" />

## Error

<Story name="Error" id="elements-inputs-textarea--error" />

## Disabled

<Story name="Disabled" id="elements-inputs-textarea--disabled" />

## Required

<Story name="Required" id="elements-inputs-textarea--required" />

## Background color

The background color of the input field can be customized by using hexadecimal color codes.

```html
<template>
  <QfcTextArea label="Name" :isRequired="true" backgroundColor="#f5f5f5" />
</template>
```

The background color act as background for textfield and the label moving up when the input is focused.
