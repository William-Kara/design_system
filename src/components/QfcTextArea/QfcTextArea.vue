<template>
  <div class="text-field">
    <textarea
      v-model="model"
      :id="id"
      :style="backgroundStyle"
      :name="label"
      :placeholder="label"
      :rows="rows"
      :disabled="isDisabled"
      :required="isRequired"
      :minlength="minLength"
      :maxlength="maxLength"
      :class="inputClasses"
    />
    <label :for="id" class="text-field__label" :style="backgroundStyle">{{ formatedLabel }}</label>
    <p class="text-field__error-message">{{ errorMessage }}</p>
  </div>
</template>

<script lang="ts" setup>
import { computed } from 'vue';
import { useVModel } from '@/utils/useVModel';

type Background = `#${string}`;
interface Props {
  id: string;
  modelValue: string;
  label: string;
  errorMessage?: string;
  minLength?: number;
  maxLength?: number;
  rows?: number;
  isRequired?: boolean;
  isDisabled?: boolean;
  backgroundColor?: Background;
}
const props = withDefaults(defineProps<Props>(), {
  modelValue: '',
  type: 'text',
  label: 'Label',
  isRequired: false,
  isDisabled: false,
  backgroundColor: '#232323',
});

const id = computed(() => {
  return props.id ? props.id : Date.now().toString(36) + Math.random().toString(36).substring(2);
});

const formatedLabel = computed(() => {
  return props.isRequired ? `${props.label}*` : props.label;
});

const emit = defineEmits(['update:modelValue']);
const model = useVModel(props, emit);

const inputClasses = computed(() => {
  return {
    'text-field__input': true,
    'is-error': props.errorMessage,
  };
});
const backgroundStyle = computed(() => {
  return { 'background-color': props.backgroundColor };
});
</script>

<style lang="scss" scoped src="./QfcTextArea.scss" />
