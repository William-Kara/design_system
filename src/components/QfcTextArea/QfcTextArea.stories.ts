import QfcTextArea from './QfcTextArea.vue';
import type { Meta, StoryObj } from '@storybook/vue3';

const meta: Meta<typeof QfcTextArea> = {
  component: QfcTextArea,
  title: 'Elements/Inputs/TextArea',
  argTypes: {
    label: {
      type: { name: 'string', required: true },
      defaultValue: 'Label',
    },
    modelValue: {
      type: { name: 'string', required: true },
      defaultValue: '',
    },
    errorMessage: {
      type: { name: 'string', required: false },
    },
    maxLength: {
      type: { name: 'number', required: false },
    },
    minLength: {
      type: { name: 'number', required: false },
    },
    isRequired: {
      type: { name: 'boolean', required: false },
      defaultValue: false,
    },
    isDisabled: {
      type: { name: 'boolean', required: false },
      defaultValue: false,
    },
    backgroundColor: {
      type: { name: 'string', required: false },
    },
  },
};
export default meta;

type Story = StoryObj<typeof meta>;
export const Primary: Story = {
  args: {
    modelValue: '',
    label: 'Label',
    errorMessage: '',
    maxLength: 255,
    minLength: 0,
    isDisabled: false,
    isRequired: false,
    backgroundColor: '#232323',
  },
};

export const Error: Story = {
  args: {
    ...Primary.args,
    errorMessage: 'Error',
  },
};

export const Disabled: Story = {
  args: {
    ...Primary.args,
    isDisabled: true,
  },
};

export const Required: Story = {
  args: {
    ...Primary.args,
    isRequired: true,
  },
};
