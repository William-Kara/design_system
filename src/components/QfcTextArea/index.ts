import QfcTextArea from './QfcTextArea.vue';

export { QfcTextArea };
export default QfcTextArea;
