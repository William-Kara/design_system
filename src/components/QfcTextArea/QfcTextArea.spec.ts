import { it, describe, beforeEach, expect } from 'vitest';

import { shallowMount } from '@vue/test-utils';
import QfcTextArea from './QfcTextArea.vue';

let wrapper: any = null;

const propsValues = {
  id: 'string',
  modelValue: 'string',
  label: 'string',
  errorMessage: 'string',
  minLength: 0,
  maxLength: 0,
  rows: 0,
  isRequired: false,
  isDisabled: false,
  backgroundColor: '#CCCCCC',
};

beforeEach(() => {
  wrapper = shallowMount(QfcTextArea, {
    propsData: propsValues,
  });
});

describe('QfcTextArea Component', () => {
  it('should mount properly', () => {
    expect(wrapper).toBeTruthy();
  });

  it('should have correct props', () => {
    expect(wrapper.props()).toEqual(propsValues);
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
