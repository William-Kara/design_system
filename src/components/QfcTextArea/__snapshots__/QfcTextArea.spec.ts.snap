// Vitest Snapshot v1

exports[`QfcTextArea Component > should render correctly 1`] = `
"<div data-v-cc2c18eb=\\"\\" class=\\"text-field\\"><textarea data-v-cc2c18eb=\\"\\" id=\\"string\\" style=\\"background-color: rgb(204, 204, 204);\\" name=\\"string\\" placeholder=\\"string\\" minlength=\\"0\\" maxlength=\\"0\\" class=\\"text-field__input is-error\\"></textarea><label data-v-cc2c18eb=\\"\\" for=\\"string\\" class=\\"text-field__label\\" style=\\"background-color: rgb(204, 204, 204);\\">string</label>
  <p data-v-cc2c18eb=\\"\\" class=\\"text-field__error-message\\">string</p>
</div>"
`;
